pkg/substitute/envsubst.go

```
@@ -70,26 +75,36 @@ func SubstituteEnvFromByte(file []byte) ([]byte, error) {
 
 // stripYamlComment takes yaml bytes and returns them back with
 // comments stripped.
-func stripYamlComment(file []byte) []byte {
-	// FIXME use better logic than regex.
-	re := regexp.MustCompile(".?#.*\n")
-	stripped := re.ReplaceAllFunc(file, func(b []byte) []byte {
-		re := regexp.MustCompile("^['\"].+['\"]|^[a-zA-Z0-9]")
-		if re.Match(b) {
-			return b
-		} else {
-			return []byte("\n")
+func stripYamlComment(file []byte) ([]byte, error) {
+	decoder := yaml.NewDecoder(bytes.NewReader(file))
+	var out bytes.Buffer
+	encoder := yaml.NewEncoder(&out)
+	for {
+		var node yaml.Node
+		err := decoder.Decode(&node)
+		if errors.Is(err, io.EOF) {
+			break
 		}
-	})
-
-	var final bytes.Buffer
-	for _, line := range bytes.Split(stripped, []byte("\n")) {
-		if len(bytes.TrimSpace(line)) > 0 {
-			final.WriteString(string(line) + "\n")
+		if err != nil {
+			return nil, err
 		}
+		removeCommentsRec(&node)
+		encoder.Encode(&node)
```

Returned error is not checked here?